### PR TITLE
[CI] Enable compatibility test matrix for unit tests

### PR DESCRIPTION
**Pull request by v1v**
## What does this PR do?

Run the compatibility unit tests in the CI on a weekly basis for the `master` branch.

On a PR basis can be enabled with:
- a GitHub comment -> `run compatibility tests` or `jenkins run the compatibility tests` or `run the compatibility tests`
- using the UI and click on the `compatibility_ci` parameter.

## Why

To get towards supporting more recent versions officially, we should add Java 12-16 to the test matrix.

## Important

`java12` is the latest java version for that flavour so, there are different ones that could be used in `jdk >12`:

- openjdk
- zulu
- adoptopenjdk

This unit test matrix will run always after the `APM ITs`, for simplicity, I didn't add more complexity in the pipeline to run more things in parallel.

## Pipeline UI

image

## Checklist

- [ ] This is an enhancement of existing features, or a new feature in existing plugins
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [ ] I have added tests that prove my fix is effective or that my feature works
  - [ ] Added an API method or config option? Document in which version this will be introduced
  - [ ] I have made corresponding changes to the documentation
- [ ] This is a bugfix
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [ ] I have added tests that would fail without this fix
- [ ] This is a new plugin
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)
  - [ ] My code follows the [style guidelines of this project](https://github.com/elastic/apm-agent-java/blob/master/CONTRIBUTING.md#java-language-formatting-guidelines)
  - [ ] I have made corresponding changes to the documentation
  - [ ] I have added tests that prove my fix is effective or that my feature works
  - [ ] New and existing [**unit** tests](https://github.com/elastic/apm-agent-java/blob/master/CONTRIBUTING.md#testing) pass locally with my changes
  - [ ] I have updated [supported-technologies.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/docs/supported-technologies.asciidoc)
  - [ ] Added an API method or config option? Document in which version this will be introduced
  - [ ] Added an instrumentation plugin? Describe how you made sure that old, non-supported versions are not instrumented by accident.
- [ ] This is something else
  - [ ] I have updated [CHANGELOG.asciidoc](https://github.com/elastic/apm-agent-java/blob/master/CHANGELOG.asciidoc)

### CODE REVIEW

**Comment by v1v** [comment]
jenkins run the matrix tests

**Comment by v1v** [comment]
jenkins run the matrix tests

**Review by cachedout — APPROVED** [review]
👍 Nice work @v1v

**Review by kuisathaverat — APPROVED** [review]

**Comment by SylvainJuge** [comment]
Good job !

Do you know if there is a way to set the stage name dynamically on the matrix axis value ?
From what I understand, the stage name is not interpolated in a declarative pipeline, thus we can't have

```
stage("${JAVA_VERSION}"){
  // ...
}
```

While not being something blocking, that would definitely be great and help spot failures, we have the Java version visible when the cursor is over the `Matrix ...`, but it's definitely not obvious.

**Review by SylvainJuge — APPROVED** [review]

**Comment by v1v** [comment]
jenkins run the matrix tests

**Comment by v1v** [comment]
> Do you know if there is a way to set the stage name dynamically on the matrix axis value ?

As far as I'm aware that's a limitation with the declarative pipeline, though, the GitHub checks should provide those details:

image

and also the GitHub comment:
image

I'll include the jdk version in the script therefore the `steps errors` in the GH comment can also show that:

image

File: Jenkinsfile [hunk 1]
```diff
@@ -305,6 +304,50 @@ pipeline {
...
+              HOME = "${env.WORKSPACE}"
+              JAVA_HOME = "${env.HUDSON_HOME}/.java/${JAVA_VERSION}"
+              PATH = "${env.JAVA_HOME}/bin:${env.PATH}"
+            }
+            steps {
+              withGithubNotify(context: "Unit Tests ${JAVA_VERSION}", tab: 'tests') {
+                deleteDir()
+                unstash 'build'
+                dir("${BASE_DIR}"){
+                  sh './mvnw test'
```
Comment (v1v) [inline Jenkinsfile:338]:
```suggestion
                  sh(label: "./mvnw test for ${JAVA_VERSION}" , script: './mvnw test')
```

**Comment by v1v** [comment]
run the matrix tests

**Comment by v1v** [comment]
Recent changes provide a better UI in the PR itself:

image

**Comment by SylvainJuge** [comment]
The matrix testing works properly, however there are plenty of tests that fail and need investigation.

Thus, if we merge it to `master` as is, it means our build will be broken for a while.
- can we split the `run as master` option and have a dedicated boolean option for those compatibility tests ?
- can we execute those tests once in a while, for example weekly so we don't get them on every build on `master` which is triggered on every PR merge ?

**Comment by SylvainJuge** [comment]
Here is the non-exhaustive list of bugs that have been found on Jdk 15 and 16:
- https://github.com/elastic/apm-agent-java/issues/1944 likely on bytebuddy side.
- https://github.com/elastic/apm-agent-java/pull/1940 Jackson Json library work-around PR
- https://github.com/elastic/apm-agent-java/issues/1942 known issue in System lambda test library, without work-around for now.

Thus as we won't be able to fix those in the next few weeks, having a separate option than the `run as main` to trigger those tests would be required.

**Comment by v1v** [comment]
> can we split the run as master option and have a dedicated boolean option for those compatibility tests ?

Yes, I'll enable a new parameter called `matrix-tests`, disabled by default

> can we execute those tests once in a while, for example weekly so we don't get them on every build on master which is triggered on every PR merge ?

Yes, I'll configure a weekly run with the `matrix-tests` parameter to be set.


Therefore, this feature will never run by default for the time being on master/tags/PRs. But still PRs can be tested with the GitHub comment (see https://github.com/elastic/apm-agent-java/pull/1915#issuecomment-888214445) or the UI parameter (`matrix-tests`)


Does it sound like a good plan?

**Comment by SylvainJuge** [comment]
Thanks @v1v , that's definitely a good plan.

We might also reuse a similar strategy for the tests on Windows, which should likely be run on a weekly basis.

**Comment by cachedout** [comment]
Looks like merging this [might have triggered the full run](https://apm-ci.elastic.co/job/apm-agent-java/job/apm-agent-java-mbp/job/master/1007/) because the parameter wasn't yet present. I'm going to re-run the master branch to ensure this is working as-expected now that the the parameter is in place.

**Comment by cachedout** [comment]
@v1v The run I kicked off appears to have still run these tests. Could[ this line the cause](https://github.com/elastic/apm-agent-java/pull/1915/files#diff-e6ffa5dc854b843b3ee3c3c28f8eae2f436c2df2b1ca299cca1fa5982e390cf8R313)?

**Pull request merged**